pkgs/development/libraries/gstreamer/good/souploader.diff

```diff
@@ -1,13 +1,164 @@
+-      c_args: [adaptive_args, soup_link_args],
+-      dependencies: [adaptive_deps, soup_link_deps],
+-      kwargs: adaptive_kwargs)
+-    adaptivedemux2_static = adaptivedemux2
+-    adaptivedemux2_shared = adaptivedemux2
+-else
+-  if default_library in ['static', 'both']
+-    # Static plugin links to libsoup directly at build time
+-    adaptivedemux2_static = static_library('gstadaptivedemux2',
+-      c_args: [adaptive_args, soup_link_args],
+-      dependencies: [adaptive_deps, soup_link_deps],
+-      kwargs: adaptive_kwargs)
+-  endif
+-  if default_library in ['shared', 'both']
+-    adaptivedemux2_shared = shared_library('gstadaptivedemux2',
+-      c_args: adaptive_args,
+-      dependencies: adaptive_deps,
+-      kwargs: adaptive_kwargs)
+-  endif
++if default_library in ['static', 'both']
++  # Static plugin links to libsoup directly at build time
++  adaptivedemux2_static = static_library('gstadaptivedemux2',
++    c_args: [adaptive_args, soup_link_args],
++    dependencies: [adaptive_deps, soup_link_deps],
++    kwargs: adaptive_kwargs)
++endif
++if default_library in ['shared', 'both']
++  adaptivedemux2_shared = shared_library('gstadaptivedemux2',
++    c_args: adaptive_args,
++    dependencies: adaptive_deps,
++    kwargs: adaptive_kwargs)
+ endif
+ 
+ # Use the static library to generate the .pc file if it's available. The shared
 diff --git a/ext/soup/gstsouploader.c b/ext/soup/gstsouploader.c
-index 85048ce303..d7d818cf95 100644
+index 9192e4dac5..500ce316bc 100644
 --- a/ext/soup/gstsouploader.c
 +++ b/ext/soup/gstsouploader.c
-@@ -181,7 +181,7 @@ gst_soup_load_library (void)
+@@ -34,12 +34,17 @@ GST_DEBUG_CATEGORY (gst_soup_debug);
+ 
+ #ifndef LINK_SOUP
+ 
+-#if defined(__APPLE__) || defined(G_OS_WIN32)
+-#error "dlopen of libsoup is only supported on Linux"
++#if defined(G_OS_WIN32)
++#error "dlopen of libsoup is only supported on Darwin and Linux"
+ #endif
+ 
++#if defined(__APPLE__)
++#define LIBSOUP_3_SONAME "libsoup-3.0.0.dylib"
++#define LIBSOUP_2_SONAME "libsoup-2.4.1.dylib"
++#else
+ #define LIBSOUP_3_SONAME "libsoup-3.0.so.0"
+ #define LIBSOUP_2_SONAME "libsoup-2.4.so.1"
++#endif
+ 
+ #define LOAD_SYMBOL(name) G_STMT_START {                                \
+     if (!g_module_symbol (module, G_STRINGIFY (name), (gpointer *) &G_PASTE (vtable->_, name))) { \
+@@ -163,16 +168,16 @@ gst_soup_load_library (void)
+     /* In order to avoid causing conflicts we detect if libsoup 2 or 3 is loaded already.
+      * If so use that. Otherwise we will try to load our own version to use preferring 3. */
+ 
+-    if ((handle = dlopen (LIBSOUP_3_SONAME, RTLD_NOW | RTLD_NOLOAD))) {
+-      libsoup_sonames[0] = LIBSOUP_3_SONAME;
+-      GST_DEBUG ("LibSoup 3 found");
+-    } else if ((handle = dlopen (LIBSOUP_2_SONAME, RTLD_NOW | RTLD_NOLOAD))) {
+-      libsoup_sonames[0] = LIBSOUP_2_SONAME;
++    if ((handle = dlopen ("@nixLibSoup2Path@/" LIBSOUP_2_SONAME, RTLD_NOW | RTLD_NOLOAD))) {
```

Comment:
Linux doesn’t use the full path for libsoup 2.4. It only uses it for libsoup3, which is duplicating the behavior of https://github.com/NixOS/nixpkgs/pull/276408.

Reply:
which libsoup the testcode is linking is irrelevant as the same issue happens when linking to libsoup_3 and then trying to dlopen libsoup_3 using the fullpath of a copy.

Reply:
In the update, I populate the trailing slash in the substitution. Should I have separate patches for both Linux and Darwin (to preserve the libsoup3 first checks for Linux)?

Reply:
i don't think switching the order is an issue as on linux libsoup_3 will get loaded using the full path if no other libsoups are detected.

to fix the subtle linux issue could do something like

```c
#ifdef __APPLE__
#define MAKE_LIB_PATH(dirname, basename) dirname basename
#else
#define MAKE_LIB_PATH(dirname, basename) basename
#endif
[...]
if ((handle = dlopen (MAKE_LIB_PATH("@nixLibSoup2Path@/", LIBSOUP_2_SONAME), RTLD_NOW | RTLD_NOLOAD))) {
```

should work so linux uses the basename and apple uses the full path.

Reply:
@paparodeo I pushed an update taking a different approach. On Darwin, it uses dyld APIs to check the loaded images for libsoup. If it finds one, it uses that. Otherwise, it defaults to libsoup3 from the store.

Linux should work exactly as it did before.

Reply:
Technically, this approach isn’t thread safe. An image could be added or removed while the search is being performed. Unfortunately, I don’t know enough about how gstreamer loads plugins whether it is safe to assume that once loaded they remain loaded, and the thread safe APIs that are available require a newer SDK than is available in nixpkgs. However, I was able to successfully use GStreamer with Wine with this patch.

Reply:
that looks good! i was playing around with `dladdr(dlsym(RTLD_DEFAULT, "soup_get_major_version"), &info)` earlier which seems to also provide the image name, though has downside of breaking if the symbol name gets changed or removed.

Reply:
> Technically, this approach isn’t thread safe. An image could be added or removed while the search is being performed. 

this seems like an unlikely event that we can live with as the current version (linux) isn't thread safe either. the image can be unloaded after detection and then will fail to load in the g_module_open call.

I don't think the gstreamer plugins will get unloaded. i thought that at init time the plugin list is searched and loaded and then that's it. tho i could be misremembering as it was 6-7 months ago since i was looking at this code.